[HDR] Disable HDR display for CSS images

shallawa opened the pull request:
#### 9d522eb64cf21c3efa3513a850797e5a9718586c
<pre>
[HDR] Disable HDR display for CSS images
<a href="https://bugs.webkit.org/show_bug.cgi?id=287985#">https://bugs.webkit.org/show_bug.cgi?id=287985#</a>
<a href="https://rdar.apple.com/145154746">rdar://145154746</a>

Reviewed by Simon Fraser.

Currently only the headroom of HTMLImageElements are checked for setting the
GraphicsLayer::drawsHDRContent(). So a layer with an HDR &lt;img&gt; element, has to
show the CSS HDR image the same way if is displayed in a layer with no HDR &lt;img&gt;
element.

For now and to be consistent with HDR CSS images, the HDR display will be
disabled for CSS images.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-css-image.html: Added.
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-css-image-expected.html: Added.
* LayoutTests/fast/images/hdr-css-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-css-image-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-css-image-expected.txt: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::canDrawHDRContent const):
(WebCore::Page::canDrawHDRContents const): Deleted.
* Source/WebCore/page/Page.h:
Rename the method canDrawHDRContents() to canDrawHDRContent() to be consistent
with  GraphicsLayer::drawsHDRContent().

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
-- BitmapImage::headroomForTesting() will be respected first if it is set and
   ImagePaintingOptions::headroom is equal to Headroom::FromImage.
-- BitmapImage::headroom() will be respected second if ImagePaintingOptions::headroom
   is equal to Headroom::FromImage.
-- Otherwise ImagePaintingOptions::headroom will be used.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
This part was missing. NativeImage can be drawn from places other than BitmapImage.

* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):

Canonical link: <a href="https://commits.webkit.org/290804@main">https://commits.webkit.org/290804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9248293597496c48784cec1a6f7e152c7e9d71bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91130 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96132 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41890 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93180 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11079 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18991 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70031 "Failure limit exceed. At least found 1 new test failure: fast/events/domactivate-sets-underlying-click-event-as-handled.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27558 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94131 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82567 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50357 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8216 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/140 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41025 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/157 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98115 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18332 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79043 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18591 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78246 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22755 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18335 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18059 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21520 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/BackgroundPainter.cpp:
```diff
@@ -485,14 +485,18 @@ void BackgroundPainter::paintFillLayer(const Color& color, const FillLayer& bgLa
             context.setDrawLuminanceMask(bgLayer.maskMode() == MaskMode::Luminance);
 
+            // FIXME: Allow HDR display for background images when CSS HDR images are able to set GraphicsLayer::drawHDRContent.
+            auto headroom = Headroom::None;
```
There's also CSS image drawing in `BorderPainter::paintBorder`, and probably more; look for calls to `paintNinePieceImage`

shallawa replied:
Fixed. The HDR display is disabled in `NinePieceImage::paint()`.

shallawa committed:
[HDR] Disable HDR display for CSS images
https://bugs.webkit.org/show_bug.cgi?id=287985#
rdar://145154746

Reviewed by Simon Fraser.

Currently only the headroom of HTMLImageElements are checked for setting the
GraphicsLayer::drawsHDRContent(). So a layer with an HDR <img> element, has to
show the CSS HDR image the same way if is displayed in a layer with no HDR <img>
element.

For now and to be consistent with HDR CSS images, the HDR display will be
disabled for CSS images.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-css-image.html: Added.
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-css-image-expected.html: Added.
* LayoutTests/fast/images/hdr-css-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-css-image-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-css-image-expected.txt: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::canDrawHDRContent const):
(WebCore::Page::canDrawHDRContents const): Deleted.
* Source/WebCore/page/Page.h:
Rename the method canDrawHDRContents() to canDrawHDRContent() to be consistent
with  GraphicsLayer::drawsHDRContent().

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
-- BitmapImage::headroomForTesting() will be respected first if it is set and
   ImagePaintingOptions::headroom is equal to Headroom::FromImage.
-- BitmapImage::headroom() will be respected second if ImagePaintingOptions::headroom
   is equal to Headroom::FromImage.
-- Otherwise ImagePaintingOptions::headroom will be used.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
This part was missing. NativeImage can be drawn from places other than BitmapImage.

* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):

Canonical link: https://commits.webkit.org/290804@main
